cmd/util/ledger/util/snapshot/payload_snapshot.go

```diff
@@ -264,3 +263,279 @@ func (p *IndexMapSnapshot) ApplyChangesAndGetNewPayloads(
+					minimaps <- result{nil, err}
+					return
+				}
+				id, err := convert.LedgerKeyToRegisterID(key)
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				minimap[id] = startIndex + i
+			}
+			minimaps <- result{minimap, nil}
+		}(start, payloads[start:end])
+
+	}
+
+	// merge the minimaps in parallel
+	pairedMinimaps := make(
+		chan struct {
+			left  result
+			right result
+		}, workers/2)
+
+	go func() {
+		// we have to pair the minimaps to merge them
+		// we have to merge a total of workers-1 times
+		// before we are left with only one map
+		numberOfPairings := workers - 1
+		for i := 0; i < numberOfPairings; i++ {
+			left := <-minimaps
+			right := <-minimaps
+			pairedMinimaps <- struct {
+				left  result
+				right result
+			}{left, right}
+		}
+		close(pairedMinimaps)
+	}()
+
+	// only half of the workers are needed to merge the maps
+	wg := sync.WaitGroup{}
+	wg.Add(workers / 2)
+	for i := 0; i < workers/2; i++ {
+		go func() {
+			defer wg.Done()
+			for pair := range pairedMinimaps {
+				if pair.left.err != nil {
+					minimaps <- result{nil, pair.left.err}
+					return
+				}
+				if pair.right.err != nil {
+					minimaps <- result{nil, pair.right.err}
+					return
+				}
+				// merge the two minimaps
+				maps.Copy(pair.left.minimap, pair.right.minimap)
+				minimaps <- result{pair.left.minimap, nil}
+			}
+		}()
+	}
+
+	// wait for all workers to finish
+	wg.Wait()
+
+	r := <-minimaps
+	if r.err != nil {
+		return nil, r.err
+	}
+
+	return r.minimap, nil
+}
+
+// createLargeMap is the same as createLargeIndexMap, but it returns a map of RegisterID to index in the payloads array
+// Generics are intentionally not used to avoid performance overhead
+func createLargeMap(
+	log zerolog.Logger,
+	payloads []*ledger.Payload,
+	workers int,
+) (map[flow.RegisterID]*ledger.Payload, error) {
+
+	workers = workersToUse(workers, payloads)
+
+	if workers == 1 {
+		r := make(map[flow.RegisterID]*ledger.Payload, len(payloads))
+		for _, payload := range payloads {
+			key, err := payload.Key()
+			if err != nil {
+				return nil, err
+			}
+			id, err := convert.LedgerKeyToRegisterID(key)
+			if err != nil {
+				return nil, err
+			}
+			r[id] = payload
+		}
+
+		return r, nil
+	}
+
+	start := time.Now()
+	log.Info().
+		Int("payloads", len(payloads)).
+		Int("workers", workers).
+		Msgf("Creating payload snapshot")
+
+	defer func() {
+		log.Info().
+			Int("payloads", len(payloads)).
+			Int("workers", workers).
+			Dur("duration", time.Since(start)).
+			Msgf("Created payload snapshot")
+	}()
+
+	type result struct {
+		minimap map[flow.RegisterID]*ledger.Payload
+		err     error
+	}
+
+	minimaps := make(chan result, workers)
+	defer close(minimaps)
+
+	// create the minimaps in parallel
+	for i := 0; i < workers; i++ {
+		start := i * len(payloads) / workers
+		end := (i + 1) * len(payloads) / workers
+		if end > len(payloads) {
+			end = len(payloads)
+		}
+
+		go func(payloads []*ledger.Payload) {
```

Comment:
@fxamacker had some feedback on the previous PR last week: https://github.com/onflow/flow-go/pull/5588#discussion_r1550044342, not sure if it sill applies

Reply:
I applied it for createLargeIndexMap. but its not relevant for createLargeMap as far as I can see.